fem/bilininteg_hcurl.cpp

```diff
@@ -4843,6 +4844,1415 @@ void MixedVectorWeakCurlIntegrator::AddMultPA(const Vector &x, Vector &y) const
+            y(local_index, e) += s;
+         }
+      }
+   });
+}
+
+static void PAHcurlApplyGradientTranspose2D(
+   const int c_dofs1D, const int o_dofs1D, const int NE,
+   const Array<double> &_B, const Array<double> &_G,
+   const Vector &_x, Vector &_y)
+{
+   auto B = Reshape(_B.Read(), c_dofs1D, c_dofs1D);
+   auto G = Reshape(_G.Read(), o_dofs1D, c_dofs1D);
+
+   auto x = Reshape(_x.Read(), 2 * c_dofs1D * o_dofs1D, NE);
+   auto y = Reshape(_y.ReadWrite(), c_dofs1D, c_dofs1D, NE);
+
+   constexpr static int MAX_D1D = HCURL_MAX_D1D;
+   MFEM_VERIFY(c_dofs1D <= MAX_D1D && o_dofs1D <= c_dofs1D, "");
+
+   MFEM_FORALL(e, NE,
+   {
+      double w[MAX_D1D][MAX_D1D];
+
+      // horizontal part (open x, closed y)
+      for (int dy = 0; dy < c_dofs1D; ++dy)
+      {
+         for (int ex = 0; ex < o_dofs1D; ++ex)
+         {
+            w[dy][ex] = 0.0;
+            for (int ey = 0; ey < c_dofs1D; ++ey)
+            {
+               const int local_index = ey*o_dofs1D + ex;
+               w[dy][ex] += B(ey, dy) * x(local_index, e);
+            }
+         }
+      }
+
+      for (int dy = 0; dy < c_dofs1D; ++dy)
+      {
+         for (int dx = 0; dx < c_dofs1D; ++dx)
+         {
+            double s = 0.0;
+            for (int ex = 0; ex < o_dofs1D; ++ex)
+            {
+               s += G(ex, dx) * w[dy][ex];
+            }
+            y(dx, dy, e) += s;
+         }
+      }
+
+      // vertical part (open y, closed x)
+      for (int dy = 0; dy < c_dofs1D; ++dy)
+      {
+         for (int ex = 0; ex < c_dofs1D; ++ex)
+         {
+            w[dy][ex] = 0.0;
+            for (int ey = 0; ey < o_dofs1D; ++ey)
+            {
+               const int local_index = c_dofs1D * o_dofs1D + ey*c_dofs1D + ex;
+               w[dy][ex] += G(ey, dy) * x(local_index, e);
+            }
+         }
+      }
+
+      for (int dy = 0; dy < c_dofs1D; ++dy)
+      {
+         for (int dx = 0; dx < c_dofs1D; ++dx)
+         {
+            double s = 0.0;
+            for (int ex = 0; ex < c_dofs1D; ++ex)
+            {
+               s += B(ex, dx) * w[dy][ex];
+            }
+            y(dx, dy, e) += s;
+         }
+      }
+   });
+}
+
+static void PAHcurlApplyGradient3D(const int c_dofs1D,
+                                   const int o_dofs1D,
+                                   const int NE,
+                                   const Array<double> &_B,
+                                   const Array<double> &_G,
+                                   const Vector &_x,
+                                   Vector &_y)
+{
+   auto B = Reshape(_B.Read(), c_dofs1D, c_dofs1D);
+   auto G = Reshape(_G.Read(), o_dofs1D, c_dofs1D);
+
+   auto x = Reshape(_x.Read(), c_dofs1D, c_dofs1D, c_dofs1D, NE);
+   auto y = Reshape(_y.ReadWrite(), (3 * c_dofs1D * c_dofs1D * o_dofs1D), NE);
+
+   constexpr static int MAX_D1D = HCURL_MAX_D1D;
+   MFEM_VERIFY(c_dofs1D <= MAX_D1D && o_dofs1D <= c_dofs1D, "");
+
+   MFEM_FORALL(e, NE,
```

Comment:
For this (and all the other kernels), I wonder if there's a way to use eg `MFEM_FORALL_2/3D` in some places or `MFEM_FOREACH_THREAD` in some of the nested for loops inside this. Would that make it faster? Has anyone who knows anything about GPUs actually looked at these kernels?

Reply:
@YohannDudouit @camierjs Please correct me if I am wrong, but my understanding is that `MFEM_FORALL_2/3D` sets up blocks of GPU threads indexed with respect to quadrature indices (`qx`, `qy`, `qz`), which is beneficial for shared memory usage. I don't see a benefit for just doing some computations on each element. The usage of `MFEM_FORALL` in this kernel assigns one GPU thread per element which should be fine. I could be wrong, or maybe my understanding needs to be updated.

Reply:
If there is fine-grain parallelism within elements then there will be benefit in assigning multiple threads (a thread block) to one element.

Reply:
Looking at this and some of the other kernel implementations more carefully, I don't think there's much to be gained here.

Reply:
For, example each entry of `w1`, `w1[dx][dy][ez]`, can be computed independently, right? I think this is the same as (or at least similar to) what all the shared memory kernels do. It is not crucial to do it right now, just something to keep in mind for the future as an optimization opportunity.

Reply:
One other observation regarding the gradient computation: in the typical use case, where `Q_k` H1 function is mapped to `Nedelec_k` H(curl) function, the 1D `B` matrix is the identity. Therefore, from a performance point of view, a specialized kernel for that case is worth implementing.

Reply:
This is a good point! I just pushed a commit that includes these specialized kernels. It works fine, but it is four new (long, lots of copied code) kernels and if we want to roll back that commit because we think the maintenance cost is higher than what we gain in performance, I would not object.